Sync v2-7-stable with v2-7-test to release 2.7.0rc1

ephraimbuddy opened the pull request:
Time for `2.7.0rc1`!

potiuk committed:
Allows to choose SSL context for SMTP provider (#33075)

* Allows to choose SSL context for SMTP provider

This change add two options to choose from when SSL SMTP connection
is created:

* default - for balance between compatibility and security
* none - in case compatibility with existing infrastructure is
  preferred

The fallback is:

* The Airflow "email", "ssl_context"
* "default"

* Update airflow/providers/smtp/CHANGELOG.rst

Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>
(cherry picked from commit e20325db38fdfdd9db423a345b13d18aab6fe578)

potiuk committed:
Allows to choose SSL context for IMAP provider (#33108)

* Allows to choose SSL context for IMAP provider

This change add two options to choose from when SSL IMAP connection is created:

* default - for balance between compatibility and security
* none - in case compatibility with existing infrastructure is preferred

The fallback is:

* The Airflow "email", "ssl_context"
* "default"

Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>
(cherry picked from commit 52ca7bfc988f4c9b608f544bc3e9524fd6564639)

ashb committed:
Correctly log when there are problems trying to chmod airflow.cfg (#33118)

Without this change log gives us a big warning/stack trace about
"TypeError: not all arguments converted during string formatting".

(cherry picked from commit af0839200565dd31787632119aec0b6a751b0ca8)

mobuchowski committed:
openlineage, bigquery: add openlineage method support for BigQueryExecuteQueryOperator (#31293)

Signed-off-by: Maciej Obuchowski <obuchowski.maciej@gmail.com>
(cherry picked from commit e10aa6ae6ad07830cbf5ec59d977654c52012c22)

potiuk committed:
aDd documentation generation for CLI commands from executors (#33081)

The #29055 moved relevant CLI command definition to executors but
it automatically removed the command from generated documentation.

This PR brings it back for both Celery and Cncf Kubernetes provider

Closes: #27932
(cherry picked from commit 879fd34e97a5343e6d2bbf3d5373831b9641b5ad)

potiuk committed:
Add notes about pre-2-7 config and cli options in providers (#33123)

The config and CLI available in provider documentation is only
used as of Airflow 2.7.0. Notes about it have been added in the
relevant files.

(cherry picked from commit 9adb265e0ee697375970c8eeb83f279a8671d91a)

mobuchowski committed:
openlineage: disable running listener if not configured (#33120)

Signed-off-by: Maciej Obuchowski <obuchowski.maciej@gmail.com>
(cherry picked from commit 11ff650e1b122aadebcea462adfae5492a76ed94)

vijayasarathib committed:
Documentation Update to enhance Readability (#32832)

* Documentation Update to enhance Readability

Update to plugin.rst to enhance readability

* Update docs/apache-airflow/authoring-and-scheduling/plugins.rst

---------

Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>
(cherry picked from commit aedefd6516101b182bc9ab3a6740ac5c0a8e2f9d)

potiuk committed:
Move openlineage configuration to provider (#33124)

(cherry picked from commit bdc10a5ff6fea0fd968345fd4a9b732be49b9761)

eladkal committed:
Clean `schedule_interval` usages from example dags (#33131)

* Clean `schedule_interval` usages from example dags

* Fix Pytest parameter name to match test argument

---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
(cherry picked from commit cdea9f176ca8f9882bd41c9c198b03421dbfc298)

potiuk committed:
Split PROD image verification to a separate step in CI (#33140)

The PROD image verification happens after the images are pulled
in the "wait for PROD images" step. This verification is pretty
helpful in detecting cases where there are some "installed airflow"
problems (for example recently it helped to avoid a circular import
problem in #33081 as one of the tests failed when images were
verified. However PROD image wait might fail for other reasons and
sometimes might be "neglected" as temporary failure.

Separating verification will allow to clearly surface that the
problem is with verification, not pulling the images.

(cherry picked from commit c31070368df5688fc44bcacf3ce5978b32dbca3d)

fritz-astronomer committed:
Add links to DAGRun / DAG / Task in templates-ref.rst (#33013)

* Revert "Revert "Add links to DAGRun / DAG / Task in templates-ref.rst (#32245)" (#33010)"

This reverts commit 9fb26db0e766d22cd67d0ed1124dd97a91b519bd.

* PR feedback

* correct class references in Templates doc

* use short name for model link

(cherry picked from commit 7dc01d8d7f3ed56f725974640dc5bd2ef7e6e807)

potiuk committed:
Add links to Celery and Kubernetes executor CLI commands to Airflow (#33143)

The #33081 added documentation generation for CLI commands contributed
by executors coming from providers. Adding Airflow documentation in
the same PR was not really possible because it created a documentation
reference circle and we had to wait until the documentation inventory
has been released in S3 and since this already happened, we can now
add links to provider CLIs from Airflow.

(cherry picked from commit 9910d92bfd1cc8ac047dbdc6f9841ab2069bc31f)

Khrol committed:
Skip served logs for non-running task try (#32561)

Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>
(cherry picked from commit 29d5e955fca5e6bee30b14ac9fcf85eebc94ae6d)

potiuk committed:
Fix flaky sqlite tests with `test_xcom_map_nest` hopefully (#33145)

Recently sqlite started to fail randomly during teardown of
`test_xcom_map_nest` or `test_xcom_map_zip_nest`. This happpened
after adding `test_xcom_map_zip_nest` . It looked very strange:

```
sqlite3.ProgrammingError: SQLite objects created in a thread can only be
used in that same thread
```

By analysing possible reasons it seems that it was a side effect
of the existing `test_xcom_map_nest` that allocated a new
session in the run method of task instance rather than pass
the sesion that is created and torrn down in the pytest fixture.

The hypothesis is that the session created in the ``test_xcom_map_nest``
were being reclaimed and closed while the `test_xcom_map_zip_nest` test
was already starting in a different thread started by Pytest.

The fix is to pass the session object to run method of the taskinstance
in the ``test_xcom_map_nest`` test.

(cherry picked from commit d1d6fc994d46aaed9c801162595cae91a1ffc19c)

potiuk committed:
Fix edge cases of "migrate/create-default-connections" (#33136)

In #32810, "airflow db migrate" command has been added and it
is used by `start-airflow` command. There were a few edge cases
not covered and this PR completes it.

* We can move the "database/load_default_connections" configuration
  to a new "deprecated" section. This is the first time we remove
  the option completely as it lost its meaning, but we likely
  still want to explain that the option was there and what it does
  when deprecated "db init" command is used.

It has no meaning when you use "airflow db migrate" or when you run the
new "airflow connections create-default-connections" commands. So we can
now remove it completely from configuraiton. It will still work
in the "airflow db init" which is deprecated, as long as we provide an
explicit fallback. Also if someone had it defined in their config or
env variable, it will continue to work even if it is not defined.

* We need to explain the change in a significant newsfragment.

* The ``start-airflow`` command supports creating default connections
  with ``--load-default-connections`` flag. This was lost after
  the change so this PR brings it back by running the new
  "airflow connections create-default-connections" command if the
  flag is used.

* The `start-airflow` breeze command can be used to start older
  versions of airflow - with ``--use-airflow-version" - those that do not
  support `airflow db migrate` command. In this case the old behaviour is
  used with setting the "AIRFLOW__DATABASE__LOAD_DEFAULT_CONNECTIONS"
  based on the flag passed and running "airflow db init" instead.

(cherry picked from commit b672ba478cdf579d159d5ddd4c823a36e606f168)

potiuk committed:
Attempt to stabilise tests for xcom_arg_map (#33150)

Similarly to #33145 - this is an attempt to stabilise flaky tests
for the test_xcom_arg_map.

Even if the mechanism is not entirely clear (provide_session should
also close the connection) seems like using pytest-fixture provided
session works better than relying on a new session created in run()
methods.

(cherry picked from commit 3dd0c999f1159a2fefbf32d9f10208a274a79a62)

potiuk committed:
Improve back-reference generation (#33149)

* better command line support with better auto-complete
* allows to run back-referance generation for individual provider packages
* better typing including using Pathlib everywhere
* handling of redirection between providers - when an operator is
  moved from one provider to the other - instead of generating invalid
  "stable" link we generate back reference for all versions of the
  provider.

(cherry picked from commit 14faf196e901e0a6fdaeb7d55e6e293b7071b501)

potiuk committed:
Upgrade ruff to latest 0.0.282 version in pre-commits (#33152)

(cherry picked from commit 6b892bf21f69a41caf670e2c498aea1c83086848)

potiuk committed:
Give the the test_xcom_ar_map test opportunity to flush the data (#33153)

Using same session in different steps of the same test has the
potential of not flishing/committing the changes between.

Seems that the #33150 traded one flakiness with another. Attempting to
make sure that the flash/commit is executed before the second run.

Error:

The test RuntimeError: number of values in row (0) differ from number of
column processors (29)

The error is strange however and indicates a bug in sqlite library.

(cherry picked from commit 6b21b79f33e245ff1612b1970d05ef692c41f15c)

potiuk committed:
Give the the test_xcom_ar_map test opportunity to flush the data (#33153)

Using same session in different steps of the same test has the
potential of not flishing/committing the changes between.

Seems that the #33150 traded one flakiness with another. Attempting to
make sure that the flash/commit is executed before the second run.

Error:

The test RuntimeError: number of values in row (0) differ from number of
column processors (29)

The error is strange however and indicates a bug in sqlite library.

(cherry picked from commit 6b21b79f33e245ff1612b1970d05ef692c41f15c)

jose-neta committed:
docs(public airflow interface): fix link (#33147)

(cherry picked from commit 763fca81ca164991d3420c0daac72ea6cf752046)

o-nikolas committed:
Configurable health check threshold for triggerer (#33089)

Recently the Triggerer was forced to a health check threshold of
`trigger_heartbeat * 2.1` with a default of 5s for the heartbeat, this
generates a threshold of 10.5s, the previous threshold was 30s, this
leads to a very unstable situation where Triggerers are not given a
reasonable amount of time to heartbeat and their triggers are taken
from them.

This change allows the user to configure this threshold the same way we
do for the scheduler.

(cherry picked from commit 6ec3b9abb67617ebbea2129421c05e45dc863bee)

AVMusorin committed:
Fix reload gunicorn workers (#32102)

* Toggle gunicorn --preload with reload_on_plugin_change

Since gunicorn can't reload a new code if starts with ``--preload``
setting, we need to check ``reload_on_plugin_change`` before set it up.

Gunicorn can't reload a new code because the code is preloaded into the
master process and worker are launched with ``fork``, they will still have
the old code.

* added warning message

* Improve warning message

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

* Improve warning message

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

* Improve warning message

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

* Minor tweak of warning message

* Mention prod issue in config description

---------

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
(cherry picked from commit f78a8363a6542a4b4c94642434424da7af7fcbbb)

Commit:
Add example for list-import-errors as a CI check (#32811)

(cherry picked from commit 1b2b666ae358f6f8839610a1026f709b1c951f23)

karakanb committed:
add dag_run_ids and task_ids filter for the batch task instance API endpoint (#32705)

* add dag_run_ids and task_ids as filter types for the batch task instance endpoint

* add version notice to the new filters

* Update the released version in OpenAPI spec

---------

Co-authored-by: pierrejeambrun <pierrejbrun@gmail.com>
(cherry picked from commit 08b1e8d749612af469625add5c7f0ad582969c39)

lzdanski committed:
Docs typo and capitalization fixes (#33176)

(cherry picked from commit d8b8202d1f3d23d75f8cd362178fae359bd641f0)

potiuk committed:
Add elasticsearch group to pre-2.7 defaults (#33166)

The elasticsearch group is likely to be moved to elasticsearch
provider. Anticipating that (see #33135) we need to move it to
pre-2.7 defaults in order to have back-compatibility for providers
that assume default values to be there.

(cherry picked from commit 2d7460450dda5cc2f20d1e8cd9ead9e4d1946909)

eumiro committed:
Remove redundant dict.keys() call (#33158)

(cherry picked from commit ba9d4e470de13031fc306772900c73b9642bb0cb)

eumiro committed:
Refactor: Simplify dict manipulation in airflow/cli (#33159)

(cherry picked from commit c074a1ecd946b2ad0f85be62aa977c269e2ffb83)

potiuk committed:
Fix installing older airflow versions in Breeze with openssl fix (#33171)

The dnspython and Flask Application Builder have a weird transitive
dependency to openssl. It does not require it - it is an optional
dependnency of email validator that uses dnspython and the dnspython
uses pyopenssl when installed. This is all fine. But when opendns
is used with older FAB < 4.1.4) and NEWER opensl (>=23.0), just
creating the FAB applicatio in flask causes an exception than
there is a missing parameter in openssl.

Since openssl is NOT declared as required dependency in neither FAB
nor dnspython, installing airflow with constraints while
having a newer version of pyopenssl does not downgrade pyopenssl
it is just mentioned in a conflicting message resulting from conflicts
with other packages already installed (such as cryptography).

This caused a problem when installin older airflow version in breeze
with --use-airflow-version switch.

Also - during checking this issue, it turned out that we were not
- by default - using the constraints for older version when installing
it in breeze (you could do that by explicitly specifying the
constraints). This was because when you used anoter way to specify
versio (commit hash, tag) it is quite complex to figure out which
constraints to use. However we can automatically derive constraints
when you specify version, or branch, which are the most common
scenarios.

This PR fixes the `--use-airflow-version` case for Airflow 2.4 and below
by:

* automatically deriving the right constraints when version or branch
  are used in `--use-airflow-version`

* adding pyopenssl specifically as dependency when
  --use-airflow-version switch is used.

* also a small quality of life improvment - when airflow db migrate
  fails (because old versio of airflow does not support it) - we inform
  the user that it happened and that we are attempting to run the
  legacy airflow db init instead.

(cherry picked from commit 5387c0c761fe161f28cee70a46009e1e67286679)

eumiro committed:
Refactor: Simplify code in models (#33181)

(cherry picked from commit 5a0494f83e8ad0e5cbf0d3dcad3022a3ea89d789)

slmg committed:
Pass app context to webserver_config.py (#32759)

(cherry picked from commit 7847b6ead3c039726bb82e0de3a39e5ef5eb00aa)

amoghrajesh committed:
Fixing typo in Dockerfile (#33180)

* Fixing typo in Dockerfile

* Fixing typo in Dockerfile

(cherry picked from commit 624cf7f75a95de63bcf7a22016b13f917669e9f2)

potiuk committed:
Fix mistakenly used "release-mangement build-docs" command (#33195)

The #32495 mistakeny replaced the "build" script with a
"release-management build-docs" command where there is no such
command - there is just "build-docs"

(cherry picked from commit 199c604d3841797b97bd4b813b36379910a532a8)

doiken committed:
Fix future DagRun rarely triggered by race conditions when max_active_runs reached its upper limit. (#31414)

* feat: select dag_model with row lock

* fix: logging that scheduling was skipped

* fix: remove unused get_dagmodel

* fix: correct log message to more generic word

---------

Co-authored-by: doiken <doiken@users.noreply.github.com>
Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>
(cherry picked from commit b53e2aeefc1714d306f93e58d211ad9d52356470)

hussein-awala committed:
Replace State by TaskInstanceState in Airflow executors (#32627)

* Replace State by TaskInstanceState in Airflow executors

* chaneg state type in change_state method, KubernetesResultsType and KubernetesWatchType to TaskInstanceState

* Fix change_state annotation in CeleryExecutor

---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
(cherry picked from commit 9556d6d5f611428ac8a3a5891647b720d4498ace)

ephraimbuddy committed:
Fix xcom view returning bytes as xcom value (#33202)

(cherry picked from commit 36c2735ca48d4c2e2d239c210d3732fd8918fed2)

potiuk committed:
Cleanup all sqlalchemy sessions before each test (#33190)

Attempt to solve some of the flakiness we saw recently - this
change will run closing all opened sqlalchemy sessions before each
test. This should be a little slower than before - especially
for the very fast tests, but it provide us with much better
isolation between the tests - thus avoiding the flakiness
that we observe recently - as documented in #33178

Hopefully this one

Fixes: #33178
(cherry picked from commit 515179fe67c0c6bd8b693c5ab313f4a0cfbacea7)

dstandish committed:
Don't ignore setups when arrowing from group (#33097)

This enables us to have a group with just setups in it.

(cherry picked from commit cd7e7bcb2310dea19f7ee946716a7c91ed610c68)

potiuk committed:
For now limit Pydantic to < 2.0.0 (#33235)

* For now limit Pydantic to < 2.0.0

Pydantic 2 adds few warnings/deprecations and until AWS dependencies
are updated, we cannot move to it anyway.

Limiting Pydantic to <2 for now should prevent the warnings and
heep our dependencies in check.

* Update setup.cfg

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

---------

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
(cherry picked from commit a752c902861d9a829d7451cc69d1343cbe5b54dc)

potiuk committed:
Fix venv detection for Python virtualenv operator (#33223)

This is a follow-up after #32939. It seems that findspec does not
cover all the cases and the previous check is also faster.

Adding check for the binary first and then falling back to spec
finding will make it faster and work in the cases where the
findspec does not work (for local development cases).

(cherry picked from commit c4fe5b8b8a3b12750bb2984aca198f7bb16b6785)

dstandish committed:
Add graph screenshots in setup-teardown howto doc (#33244)

(cherry picked from commit d45781fca224074c9e8c298816b7ff4411f9c846)

ephraimbuddy committed:
Update RELEASE_NOTES.rst

o-nikolas committed:
Don't just default to scheduler heartbeat in jobs (#33084)

(cherry picked from commit c39359eafb7da21c098c07202c2bfe155c1db9ed)

potiuk committed:
wOrkaround failing dedlock when running backfill (#32991)

The dask_executor backfill tests started to fail recently more often due
to backfill exception, and the likely cause for it is that it is now
better parallelise execution and triggering of the deadlocks because of
contention betwee dag_run state update and task state update had
become much easier.

While this PR does not fix the underlying issue, it catches the
operational error where the deadlock occured during the backfill.
and rolls back the operation.

This **should** be safe. backfil has a built-in mechanism to loop and
retry failed tasks and the test passed multiple times, completing the
backfill after this fix was applied. It was not easy to reproduce it
locally but it failed every 20-30 times. When extra logging was added,
it was always connected to OperationalException raised (and caught)
right after _per_task_process. The same exception was observed few times
when rollback was added, and despite it backfill job retried and
completed the process successfully every time. We also leave the logs
with exceptions and add reassuring messages that should make it clear
that in case backfill completes, the exceptions can be ignored as
the updates will be retried by the backfill job.

Fixes: #32778
(cherry picked from commit f616ee8dff8e6ba9b37cbce0d22235dc47d4fc93)

Usiel committed:
Enable pools to consider deferred tasks (#32709)

* Makes pools respect deferrable tasks (with extra setting)

See https://github.com/apache/airflow/discussions/21243

This commit makes pools consider deferred tasks if the `include_deferred` flag is set. By default a pool will not consider deferred tasks as occupied slots, but still show the number of deferred tasks in its stats.

---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
(cherry picked from commit 70a050b1c3a46e3eabae7f6052c95718800dc604)

bbovenzi committed:
Gantt chart: Use earliest/oldest ti dates if different than dag run start/end (#33215)

(cherry picked from commit 1dcf05f1c69e52667064259308ba5950b594b268)

potiuk committed:
Fix missing related PRs when cherry-picking (#33261)

The `airflow-github` tooling had a check if PR has been already
merged, but the PR did not work well when there was already a PR
cherry-picked that referred the PR in question in their log message.

The check was done via `git log --grep=#PR` essentially and so
it happenes that `--grep` matches not only the first line of
the message but the whole log message. So if the log message
contained reference to the cherry-picked PR it was seen as merged.

This change adds ( ) around and $ to the regexp to match not
only the PR number but also parentheses around.

(cherry picked from commit eb07618d29ce957ef4401a98f25449a23d0135ad)

dstandish committed:
Remove add_task from TaskGroup (#33262)

This was added during work for AIP-52, taking inspiration from the method on the setupteardowncontext object.  But it causes problems because it's assumed in the task_id label logic that if the group is set to prefix task ids then the task id has been prefixed.  This results in bad graph labeling in UI.  Rather than fix it now, at release time, better to revert.

(cherry picked from commit 2d2a1d699b8069f59604c6d6ea3e29853faa7945)

eumiro committed:
Refactor: Simplify code in dag_processing (#33161)

(cherry picked from commit 2f16a46db51b67e67b8406fbafe9e9add6eee235)

potiuk committed:
Change type of defualts for include_deferred (#33271)

We should change the defaults to client defaults for MSSQL compatibility
The change in #32709 was not liked by MSSQL

Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>
(cherry picked from commit d989e9dba5899e87780df9a8c5994ed1e3f8a776)

eladkal commented on RELEASE_NOTES.rst:
```diff
@@ -165,11 +166,14 @@ New Features
 Improvements
 """"""""""""
+- Configurable health check threshold for triggerer (#33089, #33084)
+- Add documentation generation for CLI commands from executors (#33081, #33143, #33081)
```
I think this can be in misc section?

ephraimbuddy replied:
Just checked now, and it's actually part of AIP-51 but not marked so, and I didn't also include AIP-51 in this changelog, I wonder which other one I'm missing? I will skip this and mark them all as part of AIP-51

eladkal commented on RELEASE_NOTES.rst:
```diff
@@ -304,7 +324,8 @@ Misc/Internal
 """""""""""""""""
-- Documentation change to enhance readability (#33006)
+- Add links to ``DAGRun / DAG / Task`` in templates-ref.rst (#33013)
+- Add example for list-import-errors as a CI check (#32811)
```
```suggestion
- Add docs of how to test for DAG Import Errors (#32811)
```

I think this explains better to users what was added

eladkal reviewed (approved):
LGTM!

pierrejeambrun commented on newsfragments/33136.significant.rst:
```diff
@@ -0,0 +1,5 @@
+The "airflow db init", "airflow db upgrade" commands and "[database] load_default_connections" configuration options are deprecated.
```
33136 added a significant newsfragment, I think we should add an entry to the `RELEASE_NOTES.rst`

ephraimbuddy replied:
Resolved. Thanks!

uranusjr committed:
Ensure include_deferred is not nullable (#33280)

* Ensure include_deferred is not nullable

This uses a multi-step process to introduce the NOT NULL constraint to
the column, needed due to MSSQL restrictions.

* MSSQL does not even support boolean literals

* Postgres does not support 0 as boolean

This is fine. SQL is fun.

* Wrap raw SQL in text()

(cherry picked from commit 4f6d5973589577ffcb7db24c82dd6708b2c00023)

vandonr-amz committed:
Add a cache to `Variable` and `Connection` when called at dag parsing time (#30259)

(cherry picked from commit dc8ad0caf4da663daf39305070813d0801ebc147)

potiuk committed:
Update version in experimental Variable cache (#33286)

Follow up after #30259

(cherry picked from commit 90fb482cdc6a6730a53a82ace49d42feb57466e5)

potiuk reviewed (approved):
Reviewed all commits. LGTM!
